src/offcanvas/offcanvas-panel.ts

```diff
@@ -0,0 +1,154 @@
+import {DOCUMENT} from '@angular/common';
+import {
+  Component,
+  ElementRef,
+  EventEmitter,
+  Inject,
+  Input,
+  NgZone,
+  OnDestroy,
+  OnInit,
+  Output,
+  ViewEncapsulation
+} from '@angular/core';
+
+import {fromEvent, Observable, Subject} from 'rxjs';
+import {filter, take, takeUntil} from 'rxjs/operators';
+
+import {getFocusableBoundaryElements} from '../util/focus-trap';
+import {Key} from '../util/key';
+import {OffcanvasDismissReasons} from './offcanvas-dismiss-reasons';
+import {ngbRunTransition, NgbTransitionOptions} from '../util/transition/ngbTransition';
+import {reflow} from '../util/util';
+
+@Component({
+  selector: 'ngb-offcanvas-panel',
+  template: '<ng-content></ng-content>',
+  encapsulation: ViewEncapsulation.None,
+  styleUrls: [],
+  host: {
+    '[class]': '"offcanvas offcanvas-" + position  + (panelClass ? " " + panelClass : "")',
+    'role': 'dialog',
+    'tabindex': '-1',
+    '[attr.aria-modal]': 'true',
+    '[attr.aria-labelledby]': 'ariaLabelledBy',
+  }
+})
+export class NgbOffcanvasPanel implements OnInit,
+    OnDestroy {
+  private _closed$ = new Subject<void>();
+  private _elWithFocus: Element | null = null;  // element that is focused prior to offcanvas opening
+
+  @Input() animation: boolean;
+  @Input() ariaLabelledBy?: string;
+  @Input() keyboard = true;
+  @Input() panelClass: string;
+  @Input() position: 'start' | 'end' | 'top' | 'bottom' = 'start';
+
+  @Output('dismiss') dismissEvent = new EventEmitter();
+
+  shown = new Subject<void>();
+  hidden = new Subject<void>();
+
+  constructor(
+      @Inject(DOCUMENT) private _document: any, private _elRef: ElementRef<HTMLElement>, private _zone: NgZone) {}
+
+  dismiss(reason): void { this.dismissEvent.emit(reason); }
+
+  ngOnInit() {
+    this._elWithFocus = this._document.activeElement;
+    this._zone.onStable.asObservable().pipe(take(1)).subscribe(() => { this._show(); });
+  }
+
+  ngOnDestroy() { this._disableEventHandling(); }
+
+  hide(): Observable<any> {
+    const {nativeElement} = this._elRef;
+    const context: NgbTransitionOptions<any> = {animation: this.animation, runningTransition: 'stop'};
+
+    const offcanvasTransition$ = ngbRunTransition(this._zone, this._elRef.nativeElement, (element) => {
+      nativeElement.classList.remove('show');
+      return () => element.style.visibility = 'hidden';
```

Comment:
why this one is needed (and L.94)? removing/adding 'show' class sets element visibility:
https://github.com/twbs/bootstrap/blob/main/scss/_offcanvas.scss#L9
https://github.com/twbs/bootstrap/blob/main/scss/_offcanvas.scss#L24

Reply:
I'm not sure if it's a workaround for a bug in scss or in bootstrap, or something else, but the css looks like this: https://github.com/twbs/bootstrap/blob/main/dist/css/bootstrap.css#L6341. 
And if you look at the Bootstrap documentation and inspect the style of the offcanvas in the demos, they do set the visible style explicitly, too. And I confirm that only adding and removing the `show` class isn't sufficient.

Reply:
Mistery solved: the CSS fix was added to Bootstrap recently: https://github.com/twbs/bootstrap/commit/0d054bb0f1484dbc0c778d3643e96e37ff46d708. Before that fix, they did indeed set the visibility explicitly.